### PR TITLE
Automated api generation

### PR DESCRIPTION
**Pull request by iLLiCiTiT**
## Description
This is proposal of how to create public api functions automatically with the same docstrings as `ServerAPI` has.

### Additional information
Added a script file `automated_api.py` which should be able to fill functions in `_api.py` based on public functions of `ServerAPI` with the same docstrings, and also fill imports of the functions in `__init__.py`. First iteration did also fill the same args, kwargs but I've hit an issue when default value was `_PLACEHOLDER` object which caused issues.

### CODE REVIEW

File: ayon_api/_api.py [hunk 1]
```diff
@@ -2035,6 +2035,49 @@ def get_folders_hierarchy(*args, **kwargs):
...
+                "ownAttrib": [],
+                "updatedAt": "2023-06-12T15:37:02.420260",
+            },
+            ...
+        ]
+
+    Args:
+        project_name (str): Project name.
+        include_attrib (Optional[bool]): Inclue attribute values
+            in output. Slower the query.
```
Comment (BigRoy) [inline ayon_api/_api.py:2071]:
```suggestion
        include_attrib (Optional[bool]): Include attribute values
            in output. Slower to query.
```

File: ayon_api/_api.py [hunk 2]
```diff
@@ -2035,6 +2035,49 @@ def get_folders_hierarchy(*args, **kwargs):
     return con.get_folders_hierarchy(*args, **kwargs)
 
 
+def get_folders_flat_hierarchy(*args, **kwargs):
```
Comment (BigRoy) [inline ayon_api/_api.py:2038]: How does this differ from just doing a `get_folders` query? Isn't `get_folders` just the flat list of all folder entities?

Comment (iLLiCiTiT) [reply ayon_api/_api.py:2038]: Well `get_folders` is using GraphQl and paging. This is predefined REST endpoint that does skip the dynamic part about fetching data which makes it faster.

Comment (BigRoy) [reply ayon_api/_api.py:2038]: That difference I did not get at all from the `get_folders_flat_hierarchy` function name nor the docstring. Also, even with that explanation I'm still not sure why they SHOULD differ. And if `GraphQl` is slower shouldn't we instead be looking into how we can optimize `get_folders` to know when not to use `GraphQL`.

Or is there still a difference between the actual returned data between the two?

Anyway, `get_folders_flat_hierarchy` - I really have no idea when or why I'd call that over `get_folders`. Yes, it's faster - sure. But having it be called `get_folders_via_rest` or whatever actually makes it clearer what the intended difference is?

Comment (iLLiCiTiT) [reply ayon_api/_api.py:2038]: > Anyway, get_folders_flat_hierarchy - I really have no idea when or why I'd call that over get_folders. Yes, it's faster - sure. But having it be called get_folders_via_rest or whatever actually makes it clearer what the intended difference is?

Intention of usage is for UI tools.

Comment (BigRoy) [reply ayon_api/_api.py]: I understand the use case for 'improvements' - I'm just saying that the function name and docstring does not clarify how it differs to me at all - and especially why it needs to differ, etc.

Comment (BigRoy) [reply ayon_api/_api.py]: It's more that `get_folders_flat_hierarchy` does not explain the difference - so then the docstring should. It's unclear which of the two to use and why.

Why couldn't `get_folders` return the optimal query using rest API instead of the filtering options of fields are not used?

Or instead we have a dedicated API interface for `rest.get_folders()` or whatever so it's clear they match, may have different arguments but it's much clearer it's making a decision to use the rest api for a reason :D

The `get_folders_flat_hierarchy` to me reads as if it will return something different than `get_folders`. Which as I understand from the conversation here - it does not?

Comment (BigRoy) [reply ayon_api/_api.py]: It's hard for me because it's still unclear to me to provide better 'ideas' than what I have in the previous comments with regards to clarifying this is REST calls only and naming the API accordingly.

Additionally still not entirely sure why these functions need to be separate? Can't we dynamically decide in `get_folders` what is the optimal query?

In essence, `get_folders_flat_hierarchy`. Not sure what makes it a 'hierarchy' call? And also not sure why then after that what makes it a "flat hierarchy" and THEN knowing somewhat how the function differs based on your explanations I'm still not sure why it's not just `get_folders` except for it using a different backend call. That would just make it exactly `get_folders` but just with a different backend call, and thus then _that_ would define the function name difference compared to `get_folders`, no?

Comment (antirotor) [reply ayon_api/_api.py]: wouldn't it be nicer if it was just an argument to `get_folders()`? Something like `get_folders(flat=True)`?

Comment (BigRoy) [reply ayon_api/_api.py]: > More of these endpoints will appear over time, usually to make something faster. GraphQl is universal and dynamic, REST endpoints are predefined.

Again, totally understand - but why not just have a `rest` interface that allows you do `rest.get_folders` because in essence the return is still just list of folder entities?

Or
```
get_folders_rest(..)  # not very nice?
get_folders_raw(..)  # not super clear, but at least clear it's a raw unchanged format compared to `get_folders`
```

I just don't see what the `flat` and `hierarchy` has anything to do with what the function actually does?

Comment (BigRoy) [reply ayon_api/_api.py]: > wouldn't it be nicer if it was just an argument to `get_folders()`? Something like `get_folders(flat=True)`?

Well, the tricky thing there is that you don't really know by signature of the function which of the arguments are actually valid for the `flat=True` call which means we'd need to be doing a lot of argument value checking, etc. - which just makes things slower. Having a dedicated `get_folders_flat` call or whatever would be faster.

I only **JUST** realized based on that comment/question that I think you're  using **flat** as terminology for what is the raw REST api query? I wouldn't have expected "flat" to mean that at all - so maybe a large part of my comments were based on that misunderstanding.

Or maybe @antirotor is misunderstanding here and `flat` doesn't actually mean that it's a REST API call. If so, all my previous points still stand.

Comment (iLLiCiTiT) [reply ayon_api/_api.py]: That's out of scope of ayon-python-api and out of my scope. I'm creating python api for ayon server, if you want to complain, then I can't help and this is not correct repository to do so.

Comment (iLLiCiTiT) [reply ayon_api/_api.py]: To be honest I think you're overthinking it as developer. Do you really need to know if the function is using graphql or rest point when you're writing custom scripts for pipeline?

The specific endpoints are for specific use-cases, but it does not mean you can't use the others. They're not legacy, the only place in ayon_core which should use this method is probably hierarchy model in tools. Other places should still use `get_folders`.

Comment (iLLiCiTiT) [reply ayon_api/_api.py]: Ok, now you understand what it does, suggestions?

Comment (BigRoy) [reply ayon_api/_api.py]: > Ok, now you understand what it does, suggestions?

I think at least remove the `hierarchy` part of the function name - because the REST endpoint used in this case also does not refer to `hierarchy` at all? If that's what you're going for to match the naming used in the rest endpoints?
Then the `flat` part in the function name is just as confusing but at least it's matching the rest endpoint more?

---

Anyway, in short, no. No new suggestions other than what I've said above. Not sure if I can bring more arguments to the table and it doesn't seem to resonate as a problem for you - and if that's the case I guess there's no problem and we'll just leave it as is and I'll just need to learn the differences.

Comment (iLLiCiTiT) [reply ayon_api/_api.py]: > I think at least remove the hierarchy part of the function name - because the REST endpoint used in this case also does not refer to hierarchy at all?

It does return all folder data required to create hierarchy of a project in UI. It is second iteration of the functionality, first was `get_folders_hierarchy` which is nesting the children. This is second iteration which skips the nesting because both pipeline tools and web frontend did convert it to flat hierarchy anyway, and the nesting made backend slower.

> Anyway, in short, no. No new suggestions other than what I've said above. Not sure if I can bring more arguments to the table and it doesn't seem to resonate as a problem for you - and if that's the case I guess there's no problem and we'll just leave it as is and I'll just need to learn the differences.

Function `get_folders` is used for something else, `get_folders_rest` should mimic what `get_folder_rest` does, which is not true either. It does resonate, but I don't have better name.

Comment (BigRoy) [reply ayon_api/_api.py]: I usually don't care too much for the function name what it's used for - but more what it does. (That's for most cases, there are cases where the method or function name is better suited to be named what it responds to, e.g. `on_update`).

Anyway, we have discussed enough - I think the issue is clear - yet I don't have a go to solution. 
We might want to track it as a separate issue so it acts as a todo or discussion towards the future to discuss the naming.

We'll leave it at that.

Comment (iLLiCiTiT) [reply ayon_api/_api.py]: I wanted to do a release, so rather now... Name `get_folders_rest` can work.

Comment (BigRoy) [reply ayon_api/_api.py]: `get_folders_rest` works for me. Then `get_folders_hierarchy` would also become `get_folders_hierarchy_rest` maybe?
Anyway, also fine with me to ignore me - I think the point came across, but I also don't have a suitable better naming that I love 100%.

Comment (iLLiCiTiT) [reply ayon_api/_api.py]: > Then get_folders_hierarchy would also become get_folders_hierarchy_rest maybe?

Is already in previous release. We would have to handle backwards compatibility.

> Since the functions already existed - maybe there's also an argument to make to 'leave it' until we know how to fix it correctly.

We can change it now without any harm as it is now only in develop. Will rename it to `get_folders_rest`.

Comment (iLLiCiTiT) [reply ayon_api/_api.py]: PR https://github.com/ynput/ayon-python-api/pull/145

File: automated_api.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    "login",
+    "logout",
+    "set_default_service_username",
+}
+EXCLUDED_IMPORT_NAMES = {"GlobalContext"}
+AUTOMATED_COMMENT = "\n".join((
+    "# ------------------------------------------------",
+    "#     This content is generated automatically.",
+    "# ------------------------------------------------",
+))
```
[the remark this answers is not in the document]
Comment (iLLiCiTiT) [reply automated_api.py:26]: It is, but I'm afraid when I see `strip` next to `#` and `-`.

[the remark this answers is not in the document]
Comment (iLLiCiTiT) [reply automated_api.py:26]: Changing.

**Review by antirotor — CHANGES_REQUESTED** [review]
This seems to be working (tested by deleting existing docstring having it regenerated).

What I would appretiate is some progress/log output from the `automated_api` - running it and having no result whatsoever in the command line is frustrating,

**Comment by antirotor** [comment]
It could also use some docstring and package level docstring explaining what it should do.

**Review by antirotor — APPROVED** [review]
